File: connectors/SLURM/Rest-Client/pkg/client/client_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

// cli/internal/client/client_test.go
package client

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestDoSetsHeaders(t *testing.T) {
	var gotName, gotToken string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotName = r.Header.Get("X-SLURM-USER-NAME")
		gotToken = r.Header.Get("X-SLURM-USER-TOKEN")
		w.WriteHeader(200)
		_, _ = w.Write([]byte(`{}`))
	}))
	defer srv.Close()

	c := New(srv.URL, "root", "tok123", "41")
	if _, err := c.do("GET", "/slurm/v0.0.41/ping", nil, nil); err != nil {
		t.Fatal(err)
	}
	if gotName != "root" {
		t.Errorf("X-SLURM-USER-NAME = %q", gotName)
	}
	if gotToken != "tok123" {
		t.Errorf("X-SLURM-USER-TOKEN = %q", gotToken)
	}
}

func TestDoUnwrapsErrors(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(400)
		_, _ = w.Write([]byte(`{"errors":[{"description":"nope","error_number":42,"error":"Bad","source":"test"}]}`))
	}))
	defer srv.Close()
	c := New(srv.URL, "root", "tok", "41")
	_, err := c.do("GET", "/x", nil, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	if !strings.Contains(err.Error(), "42") || !strings.Contains(err.Error(), "nope") {
		t.Errorf("err = %v", err)
	}
}

File: web/src/shared/api/__tests__/client.test.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { ApiError, apiFetch } from "@/shared/api/client";
import { getLastTraceId, recordTraceId } from "@/shared/api/last-trace-id";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it } from "vitest";

const TRACE_ID = "a3b1c92d3f4e5a6b7c8d9e0f12345678";

const baseHandlers = [
  http.get("*/api/v1/healthz", () =>
    HttpResponse.json({ status: "ok" }, { headers: { "x-trace-id": TRACE_ID } }),
  ),
  http.get("*/api/v1/example", () => HttpResponse.json({ ok: true })),
  http.delete("*/api/v1/grant", async ({ request }) => HttpResponse.json(await request.json())),
  http.delete("*/api/v1/users/u1/privileges/last", () =>
    HttpResponse.json(
      { error: "cannot remove the last holder of core:roles:manage" },
      { status: 409 },
    ),
  ),
];

const server = setupServer(...baseHandlers);

beforeAll(() => server.listen({ onUnhandledRequest: "error" }));
afterAll(() => server.close());
afterEach(() => server.resetHandlers(...baseHandlers));
beforeEach(() => recordTraceId(null));

describe("apiFetch", () => {
  it("prepends /api/v1 and returns the JSON body", async () => {
    const body = await apiFetch<{ status: string }>("/healthz");
    expect(body).toEqual({ status: "ok" });
  });

  it("captures X-Trace-Id into the singleton when the header is present", async () => {
    await apiFetch("/healthz");
    expect(getLastTraceId()).toBe(TRACE_ID);
  });

  it("leaves the singleton untouched when the header is absent", async () => {
    recordTraceId("previous-id");
    await apiFetch("/example");
    expect(getLastTraceId()).toBe("previous-id");
  });

  it("sends a JSON body on DELETE requests", async () => {
    const body = await apiFetch<{ reason: string }>("/grant", {
      method: "DELETE",
      body: { reason: "no longer needed" },
    });
    expect(body).toEqual({ reason: "no longer needed" });
  });

  it("derives the ApiError message from the body's error field", async () => {
    const error = await apiFetch("/users/u1/privileges/last", { method: "DELETE" }).catch(
      (e) => e,
    );
    expect(error).toBeInstanceOf(ApiError);
    expect((error as ApiError).message).toBe(
      "cannot remove the last holder of core:roles:manage",
    );
    expect((error as ApiError).status).toBe(409);
  });
});
